### DISCUSSION TITLE
Inconsitent results with qgt

### DISCUSSION THREAD

**Discussion by nkaeming** [netket/netket#1627] [Q&A] [answered]
Hi all,

currently, we are experimenting with the QGT and I found a strange behavior calculating it with the different implementations.

```python
hilbert = nk.hilbert.Spin(s=0.5, N=2)
model = nk.models.ARNNDense(hilbert=hilbert, layers=2, features=10, param_dtype=jnp.complex128)
state = nk.vqs.FullSumState(hilbert=hilbert, model=model)

qgt_dense_otf = nk.optimizer.qgt.QGTOnTheFly(state, holomorphic=True).to_dense()
qgt_dense_jac = nk.optimizer.qgt.QGTJacobianDense(state, holomorphic=True).to_dense()
print(jnp.max(jnp.abs(qgt_dense_otf - qgt_dense_jac)))
```
output: 1.2837812382952014

is this by intention or is it a bug?

Best wishes
Niklas

**Comment by gcarleo** [comment]
Can you apply it to a vector? Not sure what you are doing automatically triggers the conversion to a dense matrix

**Reply by gcarleo** [reply]
Ah yes sorry I didn't see you are converting to dense explicitly uhmm that's unexpected then

**Reply by nkaeming** [reply]
I corsschecked with an RBM. Seems to work correctly in that case.

**Reply by PhilipVinc** [reply, marked as answer]
It's because an autoregressive network is not holomorphic

**Reply by PhilipVinc** [reply]
You lie to netket, netket lies to you :P

**Reply by nkaeming** [reply]
Sure. That's true. Sorry :)

**Reply by PhilipVinc** [reply]
I recently added `nk.utils.is_probably_holomorphic` exactly to check this (am not sure if it is in a released version yet or not. If not... will soon be).
If you run it on your model you'll see it returning False.

**Reply by PhilipVinc** [reply]
And more generally, `nk.optimizer.qgt.QGTOnTheFly` cannot be converted to dense for non holomorphic objects that have complex parameters. 
The way it is computed is such that we cannot extract it. You have to resort to real parameters or `QGTJacobian` if you want a dense representation.

**Reply by nkaeming** [reply]
No worries. I just was confused and forgot to think about it.
